src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.cs

```diff
@@ -0,0 +1,557 @@
+        }
+
+        /// <devdoc>
+        ///    Gets or sets the filter string, used to determine what files are monitored in a directory.
+        /// </devdoc>
+        public string Filter
+        {
+            get
+            {
+                return _filter;
+            }
+            set
+            {
+                if (String.IsNullOrEmpty(value))
+                {
+                    // Skip the string compare for "*.*" since it has no case-insensitive representation that differs from
+                    // the case-sensitive representation.
+                    _filter = "*.*";
+                }
+                else if (String.Compare(_filter, value, StringComparison.OrdinalIgnoreCase) != 0)
+                {
+                    _filter = value;
+                }
+            }
+        }
+
+        /// <devdoc>
+        ///    Gets or sets a value indicating whether subdirectories within the specified path should be monitored.
+        /// </devdoc>
+        public bool IncludeSubdirectories
+        {
+            get
+            {
+                return _includeSubdirectories;
+            }
+            set
+            {
+                if (_includeSubdirectories != value)
+                {
+                    _includeSubdirectories = value;
+
+                    Restart();
+                }
+            }
+        }
+
+        /// <devdoc>
+        ///    Gets or sets the size of the internal buffer.
+        /// </devdoc>
+        public int InternalBufferSize
+        {
+            get
+            {
+                return _internalBufferSize;
+            }
+            set
+            {
+                if (_internalBufferSize != value)
+                {
+                    if (value < 4096)
+                    {
+                        _internalBufferSize = 4096;
+                    }
+                    else
+                    {
+                        _internalBufferSize = value;
+                    }
+
+                    Restart();
+                }
+            }
+        }
+
+        /// <summary>Allocates a buffer of the requested internal buffer size.</summary>
+        /// <returns>The allocated buffer.</returns>
+        private byte[] AllocateBuffer()
+        {
+            try
+            {
+                return new byte[_internalBufferSize];
+            }
+            catch (OutOfMemoryException)
+            {
+                throw new OutOfMemoryException(SR.Format(SR.BufferSizeTooLarge, _internalBufferSize));
+            }
+        }
+
+        /// <devdoc>
+        ///    Gets or sets the path of the directory to watch.
+        /// </devdoc>
+        public string Path
+        {
+            get
+            {
+                return _directory;
+            }
+            set
+            {
+                value = (value == null) ? string.Empty : value;
+                if (String.Compare(_directory, value, StringComparison.OrdinalIgnoreCase) != 0)
```

Comment:
Also, nit: `String.Equals` should be used here instead of `String.Compare`. See #298 for details.